### PR TITLE
CAP-0024

### PR DESCRIPTION
**Pull request by jonjove**
This PR implements CAP-0024, which introduces the new operation `PathPaymentStrictSend`.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/PathPaymentStrictSendOpFrame.cpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    }
+    innerResult().success().last =
+        SimplePaymentResult(getDestID(), getDestAsset(), maxAmountSend);
+    return true;
+}
+
+bool
+PathPaymentStrictSendOpFrame::doCheckValid(uint32_t ledgerVersion)
+{
+    // TODO(jonjove): Should destMin=0 be permitted?
```
Comment (MonsieurNicolas) [inline src/transactions/PathPaymentStrictSendOpFrame.cpp]: probably not: that way people can safely reason about price ranges

File: src/transactions/test/PathPaymentStrictSendTests.cpp [hunk 2]
```diff
@@ -0,0 +1,2285 @@
...
+        auto o3 = market.requireChangesWithOffer({}, [&] {
+            return market.addOffer(mm34, {cur4, cur3, Price{2, 1}, 10});
+        });
+
+        for_versions_from(12, *app, [&] {
+            auto actual = std::vector<ClaimOfferAtom>{};
+            market.requireChanges({{o1.key, OfferState::DELETED},
+                                   {o2a.key, OfferState::DELETED},
+                                   {o2b.key, {cur3, cur2, Price{2, 1}, 5}},
+
```
Comment (MonsieurNicolas) [inline src/transactions/test/PathPaymentStrictSendTests.cpp]: re: "0 trade"
is there a missing test case where we end up deleting 03c in this situation? (was supposed to "stay" but got adjusted to be deleted)

File: src/transactions/OfferExchange.cpp [hunk 3]
```diff
@@ -710,14 +710,26 @@ applyPriceErrorThresholds(Price price, int64_t wheatReceive, int64_t sheepSend,
             }
         }
     }
-    else
+    else if (round != RoundingType::PATH_PAYMENT_STRICT_SEND)
```
Comment (MonsieurNicolas) [inline src/transactions/OfferExchange.cpp]: maybe move to a case/switch here for clarity (one has to read the comment inside the "else" case to understand what's going on)

File: src/transactions/test/PathPaymentStrictSendTests.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonjove) [reply src/transactions/test/PathPaymentStrictSendTests.cpp]: Implemented points 1,2,4 but not sure exactly how 3 would be done without a considerable amount of duplication from OfferExchange.

File: src/transactions/OfferExchange.cpp (hunk 3 above)
Comment (jonjove) [reply src/transactions/OfferExchange.cpp]: Done.

File: src/transactions/test/PathPaymentStrictSendTests.cpp [hunk 4]
```diff
@@ -75,6 +75,56 @@ assetPathToString(const std::deque<Asset>& assets)
     }
     return r;
 };
+
+void
+checkClaimedOffers(std::vector<ClaimOfferAtom> const& actual,
```
Comment (MonsieurNicolas) [inline src/transactions/test/PathPaymentStrictSendTests.cpp]: nice!

**Comment by jonjove** [comment]
@MonsieurNicolas squashed and rebased.

**Comment by MonsieurNicolas** [comment]
r+ c7283b9e01fbf97d531581079e13fec3ad206f7c

**Pull request merged**